File: sdk/javascript/vectors/catbuffer.js
import SymbolBlockFactory from './BlockFactory.js';
import ReceiptFactory from './ReceiptFactory.js';
import NemFacade from '../src/facade/NemFacade.js';
import SymbolFacade from '../src/facade/SymbolFacade.js';
import * as nc from '../src/nem/models.js';
import * as sc from '../src/symbol/models.js';
import * as converter from '../src/utils/converter.js';
import { expect } from 'chai';
import JSONBigIntLib from 'json-bigint';
import fs from 'fs';
import path from 'path';

const JSONBigInt = JSONBigIntLib({ alwaysParseAsBig: true, useNativeBigInt: true });

describe('catbuffer vectors', () => {
	// region common test utils

	const prepareTestCases = (networkName, options) => {
		const cases = [];
		const schemasPath = path.join(process.env.SCHEMAS_PATH || '.', networkName, 'models');

		if (!fs.existsSync(schemasPath))
			throw Error(`could not find any cases because ${schemasPath} does not exist`);

		fs.readdirSync(schemasPath).filter(name => name.endsWith('.json')).forEach(name => {
			if (options && options.includes && options.includes.every(include => !name.includes(include))) {
				console.log(`skipping ${name} due to include filters`);
				return;
			}

			if (options && options.excludes && options.excludes.some(exclude => name.includes(exclude))) {
				console.log(`skipping ${name} due to exclude filters`);
				return;
			}

			const fileContent = fs.readFileSync(path.join(schemasPath, name), 'utf8');
			cases.push(...JSONBigInt.parse(fileContent));
		});

		if (0 === cases.length)
			throw Error(`could not find any cases in ${schemasPath}`);

		return cases;
	};

	const makeCamelCase = name => {
		let camelCaseName = '';
		let lastCharWasUnderscore = false;
		Array.from(name).forEach(ch => {
			if ('_' === ch) {
				lastCharWasUnderscore = true;
				return;
			}

			camelCaseName += lastCharWasUnderscore ? ch.toUpperCase() : ch;
			lastCharWasUnderscore = false;
		});

		return camelCaseName;
	};

	const isNumericNem = (key, value, type) => {
		const bigIntPropertyNames = [
			'amount', 'fee', 'rental_fee'
		];

		if (bigIntPropertyNames.some(name => key === name))
			return false;

		if ('delta' === key && 'mosaic_supply_change_transaction_v1' === type)
			return false;

		return 0xFFFFFFFFn >= value;
	};

	const isNumericSymbol = (key, value, type) => {
		const bigIntPropertyNames = [
			'amount', 'fee', 'duration', 'scoped_metadata_key', 'restriction_key',
			'new_restriction_value', 'previous_restriction_value',
			'mosaic_id', 'reference_mosaic_id', 'target_mosaic_id',
			'namespace_id', 'target_namespace_id',

			'height', 'difficulty', 'timestamp', 'harvesting_eligible_accounts_count'
		];

		if (bigIntPropertyNames.some(name => key === name))
			return false;

		if ('delta' === key && 'mosaic_supply_change_transaction_v1' === type)
			return false;

		return 0xFFFFFFFFn >= value;
	};

	const jsifyImpl = isNumeric => source => {
		const dest = {};
		Object.getOwnPropertyNames(source).forEach(key => {
			let value = source[key];

			if ('bigint' === typeof (value) && isNumeric(key, value, source.type)) {
				value = Number(value);
			} else if (Array.isArray(value)) {
				value = value.map(valueItem => {
					if ('bigint' === typeof (valueItem))
						return 'account_mosaic_restriction_transaction_v1' === source.type ? valueItem : Number(valueItem);

					if ('object' === typeof (valueItem))
						return jsifyImpl(isNumeric)(valueItem);

					return valueItem;
				});
			} else if ('object' === typeof (value) && null !== value) {
				value = jsifyImpl(isNumeric)(value);
			}

			dest[makeCamelCase(key)] = value;
		});

		return dest;
	};

	const jsifySymbol = jsifyImpl(isNumericSymbol);

	const jsifyNem = jsifyImpl(isNumericNem);

	// endregion

	// region create from descriptor

	describe('create from descriptor', () => {
		const fixupDescriptorCommon = (descriptor, module) => {
			Object.getOwnPropertyNames(descriptor).forEach(key => {
				// skip false positive due to ABC123 value that should be treated as plain string
				if ('value' === key && 'namespace_metadata_transaction_v1' === descriptor.type)
					return;

				const value = descriptor[key];
				if ('string' === typeof (value) && converter.isHexString(value))
					descriptor[key] = converter.hexToUint8(value);
				else if ('object' === typeof (value) && null !== value)
					fixupDescriptorCommon(value, module);
			});
		};

		const fixupDescriptorNem = (descriptor, module, facade) => {
			descriptor.signature = new module.Signature(descriptor.signature);
			fixupDescriptorCommon(descriptor, module);

			if (descriptor.innerTransaction) {
				fixupDescriptorNem(descriptor.innerTransaction, module, facade);
				descriptor.innerTransaction = facade.transactionFactory.constructor
					.toNonVerifiableTransaction(facade.transactionFactory.create(descriptor.innerTransaction));

				if (descriptor.cosignatures) {
					descriptor.cosignatures = descriptor.cosignatures.map(cosignatureDescriptor => {
						const cosignature = facade.transactionFactory.create({
							type: 'cosignature_v1',
							...cosignatureDescriptor.cosignature
						});
						cosignature.signature = new module.Signature(cosignatureDescriptor.cosignature.signature);
						cosignature.network = module.NetworkType.MAINNET; // TODO: fixup based on mismatch in vectors

						const sizePrefixedCosignature = new module.SizePrefixedCosignatureV1();
						sizePrefixedCosignature.cosignature = cosignature;
						return sizePrefixedCosignature;
					});
				}
			}
		};

		const fixupCosignatureSymbol = (descriptor, module) => {
			const cosignature = new module.Cosignature();
			cosignature.signature = new module.Signature(descriptor.signature);
			cosignature.signerPublicKey = new module.PublicKey(descriptor.signerPublicKey);
			return cosignature;
		};

		const fixupDescriptorSymbol = (descriptor, module, facade) => {
			descriptor.signature = new module.Signature(descriptor.signature);
			fixupDescriptorCommon(descriptor, module);

			if (descriptor.transactions) {
				descriptor.transactions = descriptor.transactions
					.map(childDescriptor => facade.transactionFactory.createEmbedded(childDescriptor));

				if (descriptor.cosignatures) {
					descriptor.cosignatures = descriptor.cosignatures
						.map(cosignatureDescriptor => fixupCosignatureSymbol(cosignatureDescriptor, module));
				}
			}

			return descriptor;
		};

		const fixupBlockDescriptorSymbol = (descriptor, module, facade) => {
			descriptor.signature = new module.Signature(descriptor.signature);
			fixupDescriptorCommon(descriptor, module);

			if (descriptor.transactions) {
				descriptor.transactions = descriptor.transactions
					.map(childDescriptor => fixupDescriptorSymbol(childDescriptor, module, facade));

				descriptor.transactions = descriptor.transactions
					.map(childDescriptor => facade.transactionFactory.create(childDescriptor));
			}
		};

		const isKeyInFormattedString = (transaction, key) => {
			if (transaction.toString().includes(key))
				return true;

			return 'parentName' === key && null === transaction[key];
		};

		const assertCreateFromDescriptor = (item, module, FacadeClass, fixupDescriptor, jsify) => {
			// Arrange:
			const facade = new FacadeClass('testnet');

			const descriptor = jsify(item.descriptor);
			fixupDescriptor(descriptor, module, facade);

			// Act:
			const transaction = facade.transactionFactory.create(descriptor);
			const transactionBuffer = transaction.serialize();

			// Assert:
			expect(converter.uint8ToHex(transactionBuffer)).to.equal(item.payload);
			expect(Object.getOwnPropertyNames(descriptor).every(key => isKeyInFormattedString(transaction, key)))
				.to.equal(true);
		};

		const createSymbolDescriptor = (originalDescriptor, fixupDescriptor) => {
			const facade = new SymbolFacade('testnet');

			// - this will be dealing with symbol blocks only
			const descriptor = jsifySymbol(originalDescriptor);
			fixupDescriptor(descriptor, sc, facade);

			return { network: facade.network, descriptor };
		};

		const assertCreateSymbolBlockFromDescriptor = (item, fixupDescriptor) => {
			// Arrange:
			const { network, descriptor } = createSymbolDescriptor(item.descriptor, fixupDescriptor);

			// Act:
			const blockFactory = new SymbolBlockFactory(network);
			const block = blockFactory.create(descriptor);
			const blockBuffer = block.serialize();

			// Assert:
			expect(converter.uint8ToHex(blockBuffer)).to.equal(item.payload);
			expect(Object.getOwnPropertyNames(descriptor).every(key => isKeyInFormattedString(block, key)))
				.to.equal(true);
		};

		const assertCreateSymbolReceiptFromDescriptor = (item, fixupDescriptor) => {
			// Arrange:
			const { descriptor } = createSymbolDescriptor(item.descriptor, fixupDescriptor);

			// Act:
			const receiptFactory = new ReceiptFactory();
			const receipt = receiptFactory.create(descriptor);
			const receiptBuffer = receipt.serialize();

			// Assert:
			expect(converter.uint8ToHex(receiptBuffer)).to.equal(item.payload);
			expect(Object.getOwnPropertyNames(descriptor).every(key => isKeyInFormattedString(receipt, key)))
				.to.equal(true);
		};

		describe('NEM', () => {
			prepareTestCases('nem').forEach(item => {
				it(`can create from descriptor ${item.test_name}`, () => {
					assertCreateFromDescriptor(item, nc, NemFacade, fixupDescriptorNem, jsifyNem);
				});
			});
		});

		describe('Symbol (transactions)', () => {
			prepareTestCases('symbol', { includes: ['transactions'] }).forEach(item => {
				it(`can create from descriptor ${item.test_name}`, () => {
					assertCreateFromDescriptor(item, sc, SymbolFacade, fixupDescriptorSymbol, jsifySymbol);
				});
			});
		});

		describe('Symbol (blocks)', () => {
			prepareTestCases('symbol', { includes: ['blocks'] }).forEach(item => {
				it(`can create from descriptor ${item.test_name}`, () => {
					assertCreateSymbolBlockFromDescriptor(item, fixupBlockDescriptorSymbol);
				});
			});
		});

		describe('Symbol (receipts)', () => {
			prepareTestCases('symbol', { includes: ['receipts'] }).forEach(item => {
				it(`can create from descriptor ${item.test_name}`, () => {
					assertCreateSymbolReceiptFromDescriptor(item, () => {});
				});
			});
		});
	});

	// endregion

	// region create from constructor

	describe('create from constructor', () => {
		const assertCreateFromConstructor = (schemaName, module) => {
			// Arrange:
			const SchemaClass = module[schemaName];

			// Act:
			const transaction = new SchemaClass();

			const { size } = transaction;
			const transactionBuffer = transaction.serialize();

			// Assert:
			expect(size).to.not.equal(0);
			expect(transactionBuffer.length).to.not.equal(0);
			expect(size).to.equal(transactionBuffer.length);
		};

		describe('NEM', () => {
			new Set(prepareTestCases('nem').map(item => item.schema_name)).forEach(schemaName => {
				it(`can create from constructor ${schemaName}`, () => {
					assertCreateFromConstructor(schemaName, nc);
				});
			});
		});

		describe('Symbol', () => {
			new Set(prepareTestCases('symbol').map(item => item.schema_name)).forEach(schemaName => {
				it(`can create from constructor ${schemaName}`, () => {
					assertCreateFromConstructor(schemaName, sc);
				});
			});
		});
	});

	// endregion

	// region roundtrip

	describe('roundtrip', () => {
		const assertRoundtrip = (item, module) => {
			// Arrange:
			const schemaName = item.schema_name;
			const payloadHex = item.payload;
			const payload = converter.hexToUint8(payloadHex);

			const TransactionClass = module[schemaName];

			// Act:
			const transaction = TransactionClass.deserialize(payload);
			const transactionBuffer = transaction.serialize();

			// Assert:
			expect(converter.uint8ToHex(transactionBuffer)).to.equal(payloadHex);
			expect(transaction.size).to.equal(transactionBuffer.length);

			if (schemaName.endsWith('Transaction'))
				assertRoundtrip({ schema_name: 'TransactionFactory', payload: payloadHex }, module);
		};

		describe('NEM', () => {
			prepareTestCases('nem').forEach(item => {
				it(`can roundtrip ${item.test_name}`, () => {
					assertRoundtrip(item, nc);
				});
			});
		});

		describe('Symbol', () => {
			prepareTestCases('symbol').forEach(item => {
				it(`can roundtrip ${item.test_name}`, () => {
					assertRoundtrip(item, sc);
				});
			});
		});
	});

	// endregion
});

File: sdk/python/tests/vectors/catbuffer.py
import importlib
import json
import os
import re
from binascii import hexlify, unhexlify
from pathlib import Path

import pytest

from symbolchain.facade.SymbolFacade import SymbolFacade
from testvectors.BlockFactory import BlockFactory as SymbolBlockFactory
from testvectors.ReceiptFactory import ReceiptFactory

# region common test utils


def prepare_test_cases(network_name, includes=None, excludes=None):
	cases = []
	schemas_path = Path(os.environ.get('SCHEMAS_PATH', '.')) / network_name / 'models'

	if not schemas_path.exists():
		raise RuntimeError(f'could not find any cases because {schemas_path} does not exist')

	for filepath in schemas_path.glob('*.json'):
		if includes and not any(include in filepath.name for include in includes):
			print(f'skipping {filepath.name} due to include filters')
			continue

		if excludes and any(exclude in filepath.name for exclude in excludes):
			print(f'skipping {filepath.name} due to exclude filters')
			continue

		with open(filepath, 'rt', encoding='utf8') as infile:
			cases += json.load(infile)

	if not cases:
		raise RuntimeError(f'could not find any cases in {schemas_path}')

	return cases


def to_hex_string(binary):
	return hexlify(binary).decode('ascii').upper()


def generate_pretty_id(val):
	return val['test_name']

# endregion


# region create from descriptor

def fixup_descriptor_common(descriptor, module):
	hex_string_pattern = re.compile('^[0-9A-F]+$')

	for key, value in descriptor.items():
		# skip false positive due to ABC123 value that should be treated as plain string
		if 'value' == key and 'namespace_metadata_transaction_v1' == descriptor.get('type'):
			continue

		if isinstance(value, str) and hex_string_pattern.match(value):
			descriptor[key] = unhexlify(value)
		elif isinstance(value, list):
			for value_item in value:
				if isinstance(value_item, dict):
					fixup_descriptor_common(value_item, module)
		elif isinstance(value, dict):
			fixup_descriptor_common(value, module)


def fixup_descriptor_nem(descriptor, module, facade):
	descriptor['signature'] = getattr(module, 'Signature')(descriptor['signature'])
	fixup_descriptor_common(descriptor, module)

	if 'inner_transaction' in descriptor:
		fixup_descriptor_nem(descriptor['inner_transaction'], module, facade)
		descriptor['inner_transaction'] = facade.transaction_factory.to_non_verifiable_transaction(
			facade.transaction_factory.create(descriptor['inner_transaction'])
		)

		if 'cosignatures' in descriptor:
			for cosignature in descriptor['cosignatures']:
				cosignature['cosignature']['signature'] = getattr(module, 'Signature')(cosignature['cosignature']['signature'])


def fixup_cosignature_symbol(descriptor, module):
	cosignature = getattr(module, 'Cosignature')()
	cosignature.signature = getattr(module, 'Signature')(descriptor['signature'])
	cosignature.signer_public_key = getattr(module, 'PublicKey')(descriptor['signer_public_key'])
	return cosignature


def fixup_descriptor_symbol(descriptor, module, facade):
	descriptor['signature'] = getattr(module, 'Signature')(descriptor['signature'])
	fixup_descriptor_common(descriptor, module)

	if 'transactions' not in descriptor:
		return

	descriptor['transactions'] = [
		facade.transaction_factory.create_embedded(child_descriptor) for child_descriptor in descriptor['transactions']
	]

	if 'cosignatures' not in descriptor:
		return

	descriptor['cosignatures'] = [
		fixup_cosignature_symbol(cosignature_descriptor, module) for cosignature_descriptor in descriptor['cosignatures']
	]


def fixup_block_descriptor_symbol(descriptor, module, facade):
	descriptor['signature'] = getattr(module, 'Signature')(descriptor['signature'])
	fixup_descriptor_common(descriptor, module)

	if 'transactions' not in descriptor:
		return

	block_transactions = []
	for block_transaction_descriptor in descriptor['transactions']:
		fixup_descriptor_symbol(block_transaction_descriptor, module, facade)
		block_transactions.append(facade.transaction_factory.create(block_transaction_descriptor))

	descriptor['transactions'] = block_transactions


def is_key_in_formatted_string(transaction, key):
	if key in str(transaction):
		return True

	return 'parent_name' == key and getattr(transaction, key) is None


def assert_create_from_descriptor(item, module, facade_name, fixup_descriptor):
	# Arrange:
	payload_hex = item['payload']

	facade_module = importlib.import_module(f'symbolchain.facade.{facade_name}')
	facade_class = getattr(facade_module, facade_name)
	facade = facade_class('testnet')

	descriptor = item['descriptor']
	fixup_descriptor(descriptor, module, facade)

	# Act:
	transaction = facade.transaction_factory.create(descriptor)
	transaction_buffer = transaction.serialize()

	# Assert:
	assert payload_hex == to_hex_string(transaction_buffer)
	assert all(is_key_in_formatted_string(transaction, key) for key in descriptor.keys())


def create_symbol_descriptor(original_descriptor, fixup_descriptor):
	facade = SymbolFacade('testnet')
	descriptor = original_descriptor
	fixup_descriptor(descriptor, importlib.import_module('symbolchain.sc'), facade)

	return facade.network, descriptor


def assert_create_symbol_block_from_descriptor(item, fixup_descriptor):  # pylint: disable=invalid-name
	# Arrange:
	network, descriptor = create_symbol_descriptor(item['descriptor'], fixup_descriptor)

	# Act:
	block = SymbolBlockFactory(network).create(descriptor)
	block_buffer = block.serialize()

	# Assert:
	assert item['payload'] == to_hex_string(block_buffer)
	assert all(is_key_in_formatted_string(block, key) for key in descriptor.keys())


def assert_create_symbol_receipt_from_descriptor(item, fixup_descriptor):  # pylint: disable=invalid-name
	# Arrange:
	_, descriptor = create_symbol_descriptor(item['descriptor'], fixup_descriptor)

	# Act:
	receipt, descriptor = ReceiptFactory().create(descriptor)
	receipt_buffer = receipt.serialize()

	# Assert:
	assert item['payload'] == to_hex_string(receipt_buffer)
	assert all(is_key_in_formatted_string(receipt, key) for key in descriptor.keys())


@pytest.mark.parametrize('item', prepare_test_cases('nem'), ids=generate_pretty_id)
def test_create_from_descriptor_nem(item):
	assert_create_from_descriptor(item, importlib.import_module('symbolchain.nc'), 'NemFacade', fixup_descriptor_nem)


@pytest.mark.parametrize('item', prepare_test_cases('symbol', includes=['transactions']), ids=generate_pretty_id)
def test_create_from_descriptor_symbol(item):  # pylint: disable=invalid-name
	assert_create_from_descriptor(item, importlib.import_module('symbolchain.sc'), 'SymbolFacade', fixup_descriptor_symbol)


@pytest.mark.parametrize('item', prepare_test_cases('symbol', includes=['blocks']), ids=generate_pretty_id)
def test_create_blocks_from_descriptor_symbol(item):  # pylint: disable=invalid-name
	assert_create_symbol_block_from_descriptor(item, fixup_block_descriptor_symbol)


def no_fixup(_1, _2, _3):
	pass


@pytest.mark.parametrize('item', prepare_test_cases('symbol', includes=['receipts']), ids=generate_pretty_id)
def test_create_receipts_from_descriptor_symbol(item):  # pylint: disable=invalid-name
	assert_create_symbol_receipt_from_descriptor(item, no_fixup)

# endregion


# region create from constructor

def assert_create_from_constructor(schema_name, module):
	# Arrange:
	schema_class = getattr(module, schema_name)

	# Act:
	transaction = schema_class()

	size = transaction.size
	transaction_buffer = transaction.serialize()

	# Assert:
	assert 0 != size
	assert 0 != len(transaction_buffer)
	assert size == len(transaction_buffer)


@pytest.mark.parametrize('item', set(test_case['schema_name'] for test_case in prepare_test_cases('nem')))
def test_create_from_constructor_nem(item):
	assert_create_from_constructor(item, importlib.import_module('symbolchain.nc'))


@pytest.mark.parametrize('item', set(test_case['schema_name'] for test_case in prepare_test_cases('symbol')))
def test_create_from_constructor_symbol(item):  # pylint: disable=invalid-name
	assert_create_from_constructor(item, importlib.import_module('symbolchain.sc'))

# endregion


# region roundtrip

def assert_roundtrip(item, module):
	# Arrange:
	schema_name = item['schema_name']
	payload_hex = item['payload']
	payload = unhexlify(payload_hex)

	transaction_class = getattr(module, schema_name)

	# Act:
	transaction = transaction_class.deserialize(payload)
	transaction_buffer = transaction.serialize()

	# Assert:
	assert payload_hex == to_hex_string(transaction_buffer)
	assert len(transaction_buffer) == transaction.size

	# - additionally pass all transactions through TransactionFactory builder ([:-2] to ignore "v1", "v2" suffix)
	if schema_name[:-2].endswith('Transaction'):
		assert_roundtrip({'schema_name': 'TransactionFactory', 'payload': payload_hex}, module)


@pytest.mark.parametrize('item', prepare_test_cases('nem'), ids=generate_pretty_id)
def test_roundtrip_nem(item):
	assert_roundtrip(item, importlib.import_module('symbolchain.nc'))


@pytest.mark.parametrize('item', prepare_test_cases('symbol'), ids=generate_pretty_id)
def test_roundtrip_symbol(item):
	assert_roundtrip(item, importlib.import_module('symbolchain.sc'))

# endregion
